File: DdcSharp/MonitorApi.cs
using System;
using System.Collections.Generic;
using System.Linq;

using DdcSharp.Models;
using DdcSharp.Native;
using DdcSharp.Native.Models;

namespace DdcSharp
{
    public class MonitorApi
    {
        // Refs
        // http://stackoverflow.com/q/846518/2001966
        // http://stackoverflow.com/a/18065609/2001966
        // Monitor functions - https://msdn.microsoft.com/en-us/library/windows/desktop/dd692964(v=vs.85).aspx

        public IEnumerable<VirtualDisplay> GetDisplays()
        {
            var monitors = new List<VirtualDisplay>();

            NativeApi.EnumDisplayMonitors(IntPtr.Zero, IntPtr.Zero,
                delegate(IntPtr hMonitor, IntPtr hdcMonitor, RECT lprcMonitor, IntPtr dwData)
                {
                    var monitorInfo = new MONITORINFOEX();
                    monitorInfo.Init();
                    var success = NativeApi.GetMonitorInfo(hMonitor, ref monitorInfo);
                    if (success)
                    {
                        var displayInfo = new VirtualDisplay
                        {
                            Width = monitorInfo.Monitor.Width,
                            Height = monitorInfo.Monitor.Height,
                            IsPrimary = monitorInfo.Flags.HasFlag(MONITORINFOEX_FLAGS.MONITORINFOF_PRIMARY),
                            Handle = hMonitor
                        };
                        monitors.Add(displayInfo);
                    }
                    return true;
                }, IntPtr.Zero);

            foreach (var displayInfo in monitors)
            {
                displayInfo.PhysicalDisplays = GetPysicalMonitors(displayInfo.Handle).ToList();

                yield return displayInfo;
            }
        }

        public IEnumerable<PhysicalDisplay> GetPysicalMonitors(IntPtr hMonitor)
        {
            uint numberOfPhysicalMonitors = 0;
            var success = NativeApi.GetNumberOfPhysicalMonitorsFromHMONITOR(hMonitor, ref numberOfPhysicalMonitors);
            if (success)
            {
                var physicalMonitors = new PHYSICAL_MONITOR[numberOfPhysicalMonitors];
                NativeApi.GetPhysicalMonitorsFromHMONITOR(hMonitor, numberOfPhysicalMonitors, physicalMonitors);
                foreach (var physicalMonitor in physicalMonitors)
                {
                    MonitorCapabilities pdwMonitorCapabilities = 0;
                    SupportedColorTemperatures pdwSupportedColorTemperatures = 0;
                    NativeApi.GetMonitorCapabilities(physicalMonitor.PhysicalMonitorHandler, ref pdwMonitorCapabilities, ref pdwSupportedColorTemperatures);

                    yield return
                        new PhysicalDisplay(
                            () => NativeApi.DestroyPhysicalMonitors(physicalMonitor.PhysicalMonitorHandler),
                            physicalMonitor.PhysicalMonitorHandler,
                            pdwMonitorCapabilities,
                            pdwSupportedColorTemperatures
                        );
                }
            }
        }
    }
}

File: DdcSharp.Core/MonitorApi.fs
namespace DdcSharp.Core

open System
open System.Collections.Generic
open NativeApis
open MonitorModels
open NativeModels
open System.Runtime.InteropServices
open System.Linq

type MonitorInfo = 
    { PHYSICAL_MONITOR : PHYSICAL_MONITOR
      MonitorCapabilities : MonitorCapabilities
      SupportedColorTemperatures : SupportedColorTemperatures }

type MonitorApi() = 
    
    member this.GetDisplays() : IEnumerable<VirtualDisplay> = 
        let monitors = new List<VirtualDisplay>()
        
        let callback (hMonitor : IntPtr) (hdcMonitor : IntPtr) (lprcMonitor : RECT) (dwData : IntPtr) = 
            let mutable monitorInfo = new MONITORINFOEX()
            monitorInfo.Size <- 104
            monitorInfo.DeviceName <- ""
            let success = GetMonitorInfo(hMonitor, &monitorInfo)
            match success with
            | true -> 
                let virtualDisplay = 
                    new VirtualDisplay(Width = monitorInfo.Monitor.Width, Height = monitorInfo.Monitor.Height, 
                                       IsPrimary = monitorInfo.Flags.HasFlag(MONITORINFOEX_FLAGS.MONITORINFOF_PRIMARY), 
                                       Handle = hMonitor, Name = monitorInfo.DeviceName)
                monitors.Add(virtualDisplay)
            | _ -> failwithf "Failed with error %i." (Marshal.GetLastWin32Error())
            true
        
        let success = EnumDisplayMonitors(IntPtr.Zero, IntPtr.Zero, new MonitorEnumDelegate(callback), IntPtr.Zero)
        if not success then failwithf "Failed with error %i." (Marshal.GetLastWin32Error())
        monitors
        |> List.ofSeq
        |> Seq.map (fun display -> 
               display.PhysicalDisplays <- this.GetPysicalMonitors(display.Handle)
               display)
    
    member this.GetPysicalMonitors(hMonitor : IntPtr) = 
        let mutable numberOfPhysicalMonitors = 0u
        let success = GetNumberOfPhysicalMonitorsFromHMONITOR(hMonitor, &numberOfPhysicalMonitors)
        if not success then failwithf "Failed with error %i." (Marshal.GetLastWin32Error())
        let physicalMonitors = 
            match success with
            | true -> 
                let mutable inMonitors = Array.zeroCreate (int32 (numberOfPhysicalMonitors))
                let success = GetPhysicalMonitorsFromHMONITOR(hMonitor, numberOfPhysicalMonitors, inMonitors)
                if not success then failwithf "Failed with error %i." (Marshal.GetLastWin32Error())
                inMonitors
            | false -> [||]
        
        let getInfo (physicalMonitor : PHYSICAL_MONITOR) = 
            let mutable pdwMonitorCapabilities = LanguagePrimitives.EnumOfValue 0u
            let mutable pdwSupportedColorTemperatures = LanguagePrimitives.EnumOfValue 0u
            let success = 
                GetMonitorCapabilities
                    (physicalMonitor.PhysicalMonitorHandler, &pdwMonitorCapabilities, &pdwSupportedColorTemperatures)
            if not success then failwithf "Failed with error %i." (Marshal.GetLastWin32Error())
            { MonitorCapabilities = pdwMonitorCapabilities
              SupportedColorTemperatures = pdwSupportedColorTemperatures
              PHYSICAL_MONITOR = physicalMonitor }
        
        let getPhysicalDisplay (info : MonitorInfo) = 
            new PhysicalDisplay(Handle = info.PHYSICAL_MONITOR.PhysicalMonitorHandler, 
                                capabilities = info.MonitorCapabilities, 
                                colorTemperatures = info.SupportedColorTemperatures, 
                                Name = info.PHYSICAL_MONITOR.PhysicalMonitorDescription)
        physicalMonitors
        |> Array.map getInfo
        |> Array.map getPhysicalDisplay
